File: src/dataStructures/MaxBinaryHeap.rb
class MaxBinaryHeap
attr_accessor :values

    def initialize
        @values = []
    end

    def insert(value) # works according to VisualAlgo
        @values.push(value)
        index = @values.length - 1
        while index > 0 do
            parent_index = ((index-1)/2).to_i
            break if @values[parent_index] > @values[index]
            @values[index] = @values[parent_index]
            @values[parent_index] = value
            index = parent_index
        end
        return self
    end

    def extract_max
        max = @values[0]
        tail = @values.pop()
        @values[0] = tail
        index = 0
        element = @values[index]
        while true do 
            swapped = nil
            left = 2 * index + 1
            right = 2 * index + 2
            if left < @values.length 
                if @values[left] > element
                    swapped = left
                end
            end
            if right < @values.length
                if (!swapped && @values[right] > element) || (!!swapped && @values[right] > @values[left])
                    swapped = right
                end
            end
            break if !swapped
            @values[index] = @values[swapped]
            @values[swapped] = element
            element = swapped
        end
        return max
    end

end

File: src/dataStructures/maxBinaryHeap.js
class MaxBinaryHeap {

    constructor() {
    this.values = []
    }

    insert(value) {
        this.values.push(value)
        let index = this.values.length - 1
        let parentIndex 
        while (!!index) {
            parentIndex = Math.floor((index-1)/2)
            if (this.values[parentIndex]>this.values[index]) break
            this.values[index] = this.values[parentIndex]
            this.values[parentIndex] = value
            index = parentIndex
        }
        return this
    }

    extractMax() {
        if (this.values.length===1) {
            return this.values.pop()
        }
        let max = this.values[0]
        let tail = this.values.pop()
        this.values[0] = tail
        let index = 0
        let element = this.values[index]
        // repeat until neither child is greater than the element
        while(true) {
            let swapped = null
        let leftIndex = 2 * index + 1
        let rightIndex = 2 * index + 2
        if (leftIndex < this.values.length) {
            if(this.values[leftIndex] > element) {
                swapped = leftIndex
            }
        }
        if (rightIndex < this.values.length) {
            if( (!swapped && this.values[rightIndex] > element || !!swapped && this.values[rightIndex] > this.values[leftIndex])) {
                swapped = rightIndex
            }
        }
        if (swapped===null) break
        this.values[index] = this.values[swapped]
        this.values[swapped] = element
        index = swapped
        }
        return max
    }

}
